### DISCUSSION TITLE
FindControl Help

### DISCUSSION THREAD

**Discussion by fischgeek** [fsprojects/Avalonia.FuncUI#293] [Q&A]
I'm trying to get a calendar to work based on the examples from Avalonia C# and the `FindControl` method. I have a feeling it has to do with some nesting issues, but can't figure it out. 

I have a TabControl inside a DockPanel in the shell.fs file where I shove the Window (this) into the **state** as "parent". So, `state.parent` contains the main window. Then, in my update method for the calendar I have tried the following but all end up not being found

```
let mainPanel = (state.parent.Content :?> DockPanel) // this actually finds the main DockPanel but everything else is not found
let tc = mainPanel.FindControl<TabControl>("navTabs") // nothing
let dates = mainPanel.FindControl<Calendar>("rangeCal") // nothing
let dates = state.parent.FindControl<Calendar>("rangeCal") // nothing
```

What am I doing wrong?

EDIT: To save on space I didn't include the shell.fs setup as it hasn't been changed other than adding the tab item and I provided `.name` to the relevant controls.

**Comment by JaggerJo** [comment]
If all you need is a reference to a (`Calendar`) control:

A. Create a view with an outlet: https://funcui.avaloniaui.net/common-questions#2.-obtain-a-view-reference-via-an-outlet

B. Use `Calendar.init (fun calendar -> ..)` to do something with the reference.

**Reply by JaggerJo** [reply]
BTW. The code you pasted probably works, but I guess you call it too early. 

```fsharp
ctx.useEffect (
    handler = (fun () -> 
        let dates = ctx.control.FindControl<Calendar>("rangeCal")
        ()
    ),
    triggers = [ EffectTrigger.AfterInit ]
)
```

**Reply by fischgeek** [reply]
> A. Create a view with an outlet: https://funcui.avaloniaui.net/common-questions#2.-obtain-a-view-reference-via-an-outlet

This looks like something I might be able to use if I knew where to put it exactly. Does this go in my elmish markup in the view method? Sorry, not sure how to use it.

> B. Use Calendar.init (fun calendar -> ..) to do something with the reference.

I want to grab the start and end dates from the user's selection. I don't think `init` would allow me to do that because it's doing it on `init` and not `onSelectedIndexChanged`?

> BTW. The code you pasted probably works, but I guess you call it too early.

I'm doing this in the update method. Is there a more appropriate place? I can also put a break point on that and traverse the controls myself, but it's like 10 levels deep and a mixture of `children` and `content`.


As for your code snippet (thank you by the way), where might that go; view? update?

Thanks for the help!

**Reply by fischgeek** [reply]
Also, `Calendar.init` does not exist it seems. 

![image](https://user-images.githubusercontent.com/4293840/230455884-5e19b6b9-dab5-4e43-b856-5bb1ec7a86f5.png)

**Reply by JaggerJo** [reply]
Which version of FuncUI + Avalonia are you on?

I highly recommend using Components (if needed with the useElmish Hook) instead of classic Elmish.

**Reply by fischgeek** [reply]
Thanks for looking at this with me. Unfortunately, I'm having more issues trying to do this task than what I believe it's worth vs just traversing controls. I tried to update to the previews, but that broke quite a few things for me. 

I'm sure somewhere along the line it's my fault and my ignorance. What was a shiny new toy for me has turned into a frustrating experience. I really want to love this, and I do in the current state using Elmish and MVU like the original template (fullmvu) offered. 

Either I take up an excessive amount of someone's time hand-holding me through this or I just deal with the limitations I have encountered so far (which I'm happy with) to get the best experience.

Thanks again for your time and help.
